Firewall: T990: Add snat and dnat connection status on firewall

nicolas-fort opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
source nat and destination nat connection status available in firewall rules configuration.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T990

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
firewall
## Proposed changes
<!--- Describe your changes in detail -->
On this first approach, we are adding only snat and dnat options for ct status.
In future PR, all options should be covered.

Netfilter references: 

- https://wiki.nftables.org/wiki-nftables/index.php/Matching_connection_tracking_stateful_metainformation#ct_status_-_conntrack_status
- https://wiki.nftables.org/wiki-nftables/index.php/Quick_reference-nftables_in_10_minutes#Ct

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
cli options
```
vyos@vyos# set firewall name FOO rule 10 
Possible completions:
   action       Rule action [REQUIRED]
 > connection-status
                Connection status
   description  Description
 > destination  Destination parameters
   disable      Option to disable firewall rule
 > fragment     IP fragment match
 > icmp         ICMP type and code information
 > ipsec        Inbound IPsec packets
 > limit        Rate limit using a token bucket filter
   log          Option to log packets matching rule
   protocol     Protocol to match (protocol name, number, or "all")
 > recent       Parameters for matching recently seen sources
 > source       Source parameters
 > state        Session state
 > tcp          TCP flags to match
 > time         Time to match rule

      
[edit]
vyos@vyos# set firewall name FOO rule 10 connection-status 
Possible completions:
   nat          NAT connection status

      
[edit]
vyos@vyos# set firewall name FOO rule 10 connection-status nat 
Possible completions:
   destination  Match connections that are subject to destination NAT
   source       Match connections that are subject to source NAT
                

```

Configurations and validations
```
set firewall ipv6-name FOO rule 10 action 'accept'
set firewall ipv6-name FOO rule 10 connection-status nat destination
set firewall ipv6-name FOO rule 10 state new 'enable'
set firewall name FOO rule 10 action 'drop'
set firewall name FOO rule 10 connection-status nat source
set firewall name FOO rule 10 state invalid 'enable'

vyos@vyos# sudo nft list chain ip filter NAME_FOO
table ip filter {
        chain NAME_FOO {
                ct state { invalid } ct status { snat } counter packets 0 bytes 0 drop comment "FOO-10"
                counter packets 0 bytes 0 return comment "FOO default-action accept"
        }
}
[edit]
vyos@vyos# sudo nft list chain ip6 filter NAME6_FOO
table ip6 filter {
        chain NAME6_FOO {
                ct state { new } ct status { dnat } counter packets 0 bytes 0 return comment "FOO-10"
                counter packets 0 bytes 0 return comment "FOO default-action accept"
        }
}
```
## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

nicolas-fort committed:
Firewall: T990: Add snat and dnat connection status on firewall

dmbaturin commented on interface-definitions/include/firewall/common-rule.xml.i:
```diff
@@ -95,6 +95,51 @@
+      <properties>
+        <help>Set when connection needs DNAT in original direction</help>
+        <completionHelp>
+          <list>enable disable</list>
```
I understand that you added "enable/disable" for uniformity with the `state` node, so I can't blame you. But our general agreement was that boolean flags should be made valueless (i.e. just `set ct-status dnat`), since values like "enable/disable" or "true/false" don't add any information to the config. They also make code more complicated since you need to handle the case when a node is absent and also the case when it's set to "disable".

I believe we should apply that policy here and make it valueless, since eventually the `state` node will also be converted to valueless. A little temporary non-uniformity is an acceptable cost for less migration in the future.

dmbaturin commented on interface-definitions/include/firewall/common-rule.xml.i:
```diff
@@ -95,6 +95,51 @@
+  <children>
+    <leafNode name="dnat">
+      <properties>
+        <help>Set when connection needs DNAT in original direction</help>
```
I think a better wording would be `Match connections that are subject to destination NAT`. The word "set" is clear in the original context in the nftables flag descriptions, but in our CLI I feel it would be ambiguous and may be interpreted as if this option _sets_ a flag, rather than make the rule match connections where that flag is _already set_.

dmbaturin commented on interface-definitions/include/firewall/common-rule.xml.i:
```diff
@@ -95,6 +95,51 @@
     </constraint>
   </properties>
 </leafNode>
+<node name="ct-status">
```
I believe `connection-status` would be a better name. Experienced Linux network admins will associate `ct` with conntrack immediately, but many VyOS users actually come from non-Linux background and that name will be opaque to them.

dmbaturin commented on interface-definitions/include/firewall/common-rule.xml.i:
```diff
@@ -95,6 +95,51 @@
 </leafNode>
+<node name="ct-status">
+  <properties>
+    <help>Connection status in conntrack</help>
```
Maybe just "Connection status". Conntrack is an implementation detail that many users may not really care about, or even know what conntrack is.

dmbaturin reviewed (changes requested):
I think the idea and implementation are good, but the CLI command names and help strings can be improved, I left some suggestions there.

c-po commented:
snat and dnat are common terms - VyOS refers to it as destination nat and source nat on the CLI:

`set nat source | destination` maybe we can create an analogy here?

`set firewall name FOO rule 10 conntrack nat source`
`set firewall name FOO rule 10 conntrack nat destination`

?

dmbaturin commented:
@c-po I'm not fond of exposing internal names in the user-visible CLI. We have changed underlying components quite a few times already, and had components change their names in the upstream (through direct renaming or a fork).

If we named the firewall "set iptables",  it would have become an anachronism and a misnomer after migration to nftables. There's a small but non-zero chance that kernel developers will replace conntrack with something else (hypothetically, "nftrack") and we'll have to either keep a name that became wrong or make a syntax change.

I think `connection-status` is both neutral and clear, but if you have better ideas, please share!

nicolas-fort commented:
Daniil.
Yes, as you said, I added "enable/disable" for uniformity with the `state` node. I'll will remove `enable/disable` for this new feature.
So, if everyone agrees, new command should be `set firewall name <name> rule <rule> connection-status [snat|dnat]`

Of course, with corrected info on helpers/descriptions.

nicolas-fort committed:
Firewall: T990: Add snat and dst connection status on firewall

c-po commented:
> @c-po I'm not fond of exposing internal names in the user-visible CLI. We have changed underlying components quite a few times already, and had components change their names in the upstream (through direct renaming or a fork).
> 
> If we named the firewall "set iptables", it would have become an anachronism and a misnomer after migration to nftables. There's a small but non-zero chance that kernel developers will replace conntrack with something else (hypothetically, "nftrack") and we'll have to either keep a name that became wrong or make a syntax change.
> 
> I think `connection-status` is both neutral and clear, but if you have better ideas, please share!

I took a very bad example here. My main issue is the snat and dnat terminus which we should name nat source and nat destination to match the NAT subsystem.

`set firewall name <name> rule <rule> connection-status nat <source | destination >` would work for me.

nicolas-fort commented:
Christians proposal sound good to me.
@dmbaturin? I already made the changes you requested. So, we leave it as it is, or move to 
`set firewall name <name> rule <rule> connection-status nat <source | destination >`??

nicolas-fort committed:
Firewall: T990: Modifications for new connection-status cli

nicolas-fort commented:
> > @c-po I'm not fond of exposing internal names in the user-visible CLI. We have changed underlying components quite a few times already, and had components change their names in the upstream (through direct renaming or a fork).
> > If we named the firewall "set iptables", it would have become an anachronism and a misnomer after migration to nftables. There's a small but non-zero chance that kernel developers will replace conntrack with something else (hypothetically, "nftrack") and we'll have to either keep a name that became wrong or make a syntax change.
> > I think `connection-status` is both neutral and clear, but if you have better ideas, please share!
> 
> I took a very bad example here. My main issue is the snat and dnat terminus which we should name nat source and nat destination to match the NAT subsystem.
> 
> `set firewall name <name> rule <rule> connection-status nat <source | destination >` would work for me.

Sound good to me too. Already changed and updated.

c-po commented:
This implementation supports either source or destination nat. A rule can not have both source and destination, as the leafNode is not a multi node.

Is this intentional?
